pandas/tests/test_multilevel.py

```diff
@@ -1490,6 +1490,19 @@ def test_frame_dict_constructor_empty_series(self):
         DataFrame({"foo": s1, "bar": s2, "baz": s3})
         DataFrame.from_dict({"foo": s1, "baz": s3, "bar": s2})
 
+    def test_empty_frame_groupby_dtypes_consistency(self):
+        # GH 20888
+        group_keys = ["a", "b", "c"]
+        df = DataFrame({"a": [1], "b": [2], "c": [3], "d": [4]})
+        g = df[df.a == 2].groupby(group_keys)
+        expected = g.first().index
+
+        df = DataFrame({"a": [1], "b": [2], "c": [3], "d": ["d"]})
+        g = df[df.a == 2].groupby(group_keys)
+        result = g.first().index
+
+        tm.assert_index_equal(result, expected)
```

Comment:
Actually, let's do a parameterization over `d` as follows:

~~~python
def test_empty_frame_groupby_dtypes_consistency(self, d):
   group_keys = [...]
   df = DataFrame({"a": [1], "b": [2], "c": [3], "d": [d]})
   g = df[df.a == 2].groupby(group_keys)

   expected = ...  # Construct the expected result
   result = g.first().index

   tm.assert_index_equal(result, expected)
~~~

Reply:
Thanks for the feedback, @gfyoung! Changes done.